firebase-firestore/src/main/java/com/google/firebase/firestore/util/ListenerRegistrationImpl.java

```diff
@@ -33,6 +35,115 @@
+  private static final String FRAGMENT_TAG = "FirestoreOnStopObserverFragment";
+
+  private void onActivityStopCallOnce(Activity activity, Runnable callback) {
+    // Android provides lifecycle callbacks (eg onStop()) that custom `Activity`s can extend. But we
+    // can't take advantage of that, since we need to be usable with a generic Activity. So instead,
+    // we create a custom Fragment, and add that Fragment to the given Activity. When the Activity
+    // stops, it will automatically stop the attached Fragments too.
+    //
+    // One difficulty with this approach is that how you get a Fragment and attach it to an Activity
+    // differs based on the type of Activity. If the Activity is actually a FragmentActivity, then
+    // you must use the android.support.v4.app.FragmentManager to do so. Otherwise, you need to use
+    // the deprecated android.app.FragmentManager.
+    //
+    // Possible improvements:
+    // 1) Allow other lifecycle callbacks other than just 'onStop'.
+    // 2) Use LifecycleOwner (which FragmentActivity implements, but Activity does not) to register
+    //    for lifecycle callbacks instead of creating/attaching a Fragment.
+
+    if (activity instanceof FragmentActivity) {
+      FragmentActivity fragmentActivity = (FragmentActivity) activity;
+
+      StopListenerSupportFragment fragment = null;
+      try {
+        fragment =
+            (StopListenerSupportFragment)
+                fragmentActivity
+                    .getSupportFragmentManager()
+                    .findFragmentByTag(SUPPORT_FRAGMENT_TAG);
+      } catch (ClassCastException e) {
+        throw new IllegalStateException(
+            "Fragment with tag '" + SUPPORT_FRAGMENT_TAG + "' is not a StopListenerSupportFragment",
+            e);
+      }
+
+      if (fragment == null || fragment.isRemoving()) {
+        fragment = new StopListenerSupportFragment();
+        fragmentActivity
+            .getSupportFragmentManager()
+            .beginTransaction()
+            .add(fragment, SUPPORT_FRAGMENT_TAG)
+            .commitAllowingStateLoss();
+      }
+
+      fragment.addCallback(callback);
+
+    } else {
+      // If we get here, then we have a non-FragmentActivity Activity. Unfortunately, all Fragment
+      // related classes/methods with non-FragmentActivity Activity's are deprecated, implying that
+      // almost everything in this block is deprecated.
+
+      StopListenerFragment fragment = null;
+      try {
+        fragment =
+            (StopListenerFragment) activity.getFragmentManager().findFragmentByTag(FRAGMENT_TAG);
+      } catch (ClassCastException e) {
+        throw new IllegalStateException(
+            "Fragment with tag '" + FRAGMENT_TAG + "' is not a StopListenerFragment", e);
+      }
+
+      if (fragment == null || fragment.isRemoving()) {
+        fragment = new StopListenerFragment();
+        activity
+            .getFragmentManager()
+            .beginTransaction()
```

Comment:
If these methods actually are deprecated we should add the suppressions to ensure they don't generate warnings.